### ISSUE TITLE
checkers/dupSubExpr_checker.go - Clarification / false positive?

### ISSUE THREAD

**Issue by dertseha** [go-critic/go-critic#897]
Several lines in my code are marked by `dupSubExpr` as 
```
dupSubExpr: suspicious identical LHS and RHS for `==` operator (gocritic)
```
Two examples:

```
func (cmd DrawCommand) TextureID() TextureID {
	var id C.IggTextureID
	C.iggDrawCommandGetTextureID(cmd.handle(), &id) // <- this line
	return TextureID(id)
}

func (data DrawData) CommandLists() []DrawList {
	var handles unsafe.Pointer
	var count C.int

	C.iggDrawDataGetCommandLists(data.handle(), &handles, &count) // <- this line
        ...
```

This can be found by running `golangci-lint` on https://github.com/inkyblackness/imgui-go on [this commit](https://github.com/inkyblackness/imgui-go/commit/632cf766904e52c23b4c6f51f70646c1a2fa36c4) (later ones will have the checker disabled).

The only common things of the marked lines were:
* All calls to C function, although not all C function calls in the library are marked (by far)
* None of the lines have the reported `==` in use

I don't see duplicated sub-expressions, and based on the non-existent `==` operator, I am tending towards a false positive.

Versions in use:
```
> golangci-lint --version
golangci-lint has version 1.21.0 built from 645e794 on 2019-10-15T18:16:56Z
> go version
go version go1.13.4 linux/amd64
```

**Comment by dominikh** [comment]
This is a false positive, caused by a peculiarity in cgo. Given the following input

```
package pkg

// void foo(void **p) {}
import "C"
import "unsafe"

func Foo() {
	var p unsafe.Pointer

	C.foo(&p)
}
```

`go tool cgo` will rewrite the code to read as such:

```
// Code generated by cmd/cgo; DO NOT EDIT.

//line foo.go:1:1
package pkg

// void foo(void **p) {}
import _ "unsafe"
import "unsafe"

func Foo() {
	var p unsafe.Pointer

	func() { _cgoBase0 := /*line :10:8*/&p; _cgo0 := _cgoBase0; _cgoCheckPointer(_cgoBase0, 0 == 0); _Cfunc_foo(_cgo0); }()
}
```

Note the `0 == 0` in the call to _cgoCheckPointer. `0 == 0` is used instead of `true` to avoid a rare edge case where the identifier `true` might be shadowed. It's this comparison that go-critic is flagging.

IIRC this comparison is used in places other than calls to `_cgoCheckPointer`, too, but someone else will have to confirm that.

**Comment by quasilyte** [comment]
If you're using `go-critic` from `golangci-lint`, it could be an option to switch `dupSubExpr` off, since `staticcheck` can detect this and the only case in the past where there was a difference is expressions inside the parenthesis, like `(x) == x`.

**Comment by dertseha** [comment]
yup, disabling this one checker is what I did :)
Thank you both for clarifying!
